src/lib/communication/DY.Lib.Communication.RequestResponse.Invariants.fst

```diff
@@ -140,48 +174,49 @@ let default_comm_reqres_higher_layer_event_preds (a:Type) {| parseable_serialize
-      parse_and_pred (fun request -> higher_layer_resreq_preds.send_request tr client server request (get_label tr key)) request
+      key `has_usage tr` (AeadKey (comm_layer_aead_tag a) empty) /\
+      higher_layer_resreq_preds.send_request tr client server request (get_label tr key)
     )
     | CommServerReceiveRequest server request key -> (
       is_knowable_by (principal_label server) tr key /\
-      is_knowable_by (get_label tr key) tr request /\
-      key `has_usage tr` (AeadKey comm_layer_aead_tag empty) /\
+      is_well_formed a (is_knowable_by (get_label tr key) tr) request /\
+      key `has_usage tr` (AeadKey (comm_layer_aead_tag a) empty) /\
       (
-        (exists client. event_triggered tr client (CommClientSendRequest client server request key)) \/
+        (exists client. event_triggered tr client (CommClientSendRequest client server request key <: communication_reqres_event a)) \/
         is_publishable tr key
       )
     )
-    | CommServerSendResponse server request response -> (
-      match parse a request, parse a response with
-      | Some request, Some response -> higher_layer_resreq_preds.send_response tr server request response
-      | _ -> False
-    )
+    | CommServerSendResponse server request response -> higher_layer_resreq_preds.send_response tr server request response
     | CommClientReceiveResponse client server response key -> (
-      (exists request. event_triggered tr server (CommServerSendResponse server request response)) \/
+      (exists request. event_triggered tr server (CommServerSendResponse server request response <: communication_reqres_event a)) \/
       is_corrupt tr (principal_label client) \/ is_corrupt tr (principal_label server)
     )
     )
 #pop-options
 
 // Additional event preconditions for the events from the core communication layer
 #push-options "--fuel 0 --ifuel 2"
-val request_response_event_preconditions: {|cinvs:crypto_invariants|} -> comm_higher_layer_event_preds com_message_t
-let request_response_event_preconditions #cinvs = {
-  (default_comm_higher_layer_event_preds com_message_t) with
-  send_conf = (fun tr client server (com_msg_t:com_message_t) ->
+val comm_core_higher_layer_event_preds_reqres:
+  {|cinvs:crypto_invariants|} ->
+  (a:Type0) -> {|comm_layer_reqres_config a|} ->
+  comm_core_higher_layer_event_preds comm_message_t #(comm_layer_tag_core_config_reqres a)
+let comm_core_higher_layer_event_preds_reqres #cinvs a #config = {
+  (default_comm_core_higher_layer_event_preds comm_message_t #(comm_layer_tag_core_config_reqres a)) with
+  send_conf = (fun tr client server (com_msg_t:comm_message_t) ->
     match com_msg_t with
     | RequestMessage {request; key} -> (
-      event_triggered tr client (CommClientSendRequest client server request key)
+      match parse a request with
+      | None -> False
+      | Some request_parsed -> event_triggered tr client (CommClientSendRequest client server request_parsed key <: communication_reqres_event a)
```

Comment:
As before with `parse_and_pred` --- this comes up several more places too and should probably be handled uniformly.